### PR TITLE
Promote WireType to the runtime.

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: wire-runtime/src/main/java/com/squareup/wire/WireType.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    scalarTypes.put(SFIXED64.string, SFIXED64);
+    scalarTypes.put(SINT32.string, SINT32);
+    scalarTypes.put(SINT64.string, SINT64);
+    scalarTypes.put(STRING.string, STRING);
+    scalarTypes.put(UINT32.string, UINT32);
+    scalarTypes.put(UINT64.string, UINT64);
+    SCALAR_TYPES = Collections.unmodifiableMap(scalarTypes);
+  }
+
+  private final boolean isScalar;
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/WireType.java]: This no longer keeps a list of nested names. It also doesn't keep a separate package name.

File: wire-runtime/src/main/java/com/squareup/wire/WireType.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  public boolean isScalar() {
+    return isScalar;
+  }
+
+  public static WireType get(String packageName, String typeName) {
+    return packageName != null ? get(packageName + '.' + typeName) : get(typeName);
+  }
+  public static WireType get(String name) {
```
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/WireType.java]: nit: missing empty line

File: wire-runtime/src/main/java/com/squareup/wire/WireType.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ */
+package com.squareup.wire;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Names a protocol buffer message, enumerated type, service or a scalar. This class models a
```
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/WireType.java]: nit: missing oxford :open_mouth:

File: wire-runtime/src/test/java/com/squareup/wire/WireTypeTest.java [hunk 4]
```diff
@@ -13,15 +13,24 @@
...
 import org.junit.Test;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.fail;
 
 public final class WireTypeTest {
-  @Test public void getScalar() throws Exception {
-    assertThat(WireType.getScalar("int32")).isSameAs(WireType.INT32);
+  @Test public void get() throws Exception {
+    assertThat(WireType.get("int32")).isSameAs(WireType.INT32);
```
Comment (JakeWharton) [inline wire-runtime/src/test/java/com/squareup/wire/WireTypeTest.java]: interesting!

File: wire-runtime/src/main/java/com/squareup/wire/WireType.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  public String enclosingTypeOrPackage() {
+    int dot = string.lastIndexOf('.');
+    return dot == -1 ? null : string.substring(0, dot);
+  }
+
+  public boolean isScalar() {
+    return isScalar;
+  }
+
+  public static WireType get(String packageName, String typeName) {
```
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/WireType.java]: nit: technically should be named `enclosingTypeOrPackage`

**Comment by JakeWharton** [comment]
What's the motivation here since it's not actually used in the runtime?

**Comment by swankjesse** [comment]
Well, I'm going to use it in the runtime! Message.Datatype is on the chopping block.

File: wire-runtime/src/main/java/com/squareup/wire/WireType.java (hunk 5 above)
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/WireType.java]: Fixed!

File: wire-runtime/src/main/java/com/squareup/wire/WireType.java (hunk 3 above)
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/WireType.java]: Fixed!

File: wire-runtime/src/main/java/com/squareup/wire/WireType.java (hunk 2 above)
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/WireType.java]: Fixed!

**Comment by JakeWharton** [comment]
I figured, just wasn't clear on where. Shared model FTW.

**Pull request merged**